### PR TITLE
Refactoring & Fixes

### LINKED ISSUE #5: Cleaner checking of Chinese charset
Maybe this would be easier to maintain when [checking chinese characterset](https://github.com/mjpost/sacreBLEU/blob/master/sacrebleu.py#L366), the core function don't  change if we were to add more or remove from the charset: 

```
chinese_charset = [
(u'\u3400', u'\u4db5'),  # CJK Unified Ideographs Extension A, release 3.0
(u'\u4e00', u'\u9fa5'),  # CJK Unified Ideographs, release 1.1
(u'\u9fa6', u'\u9fbb'),  # CJK Unified Ideographs, release 4.1
(u'\uf900', u'\ufa2d'),  # CJK Compatibility Ideographs, release 1.1
(u'\ufa30', u'\ufa6a'),  # CJK Compatibility Ideographs, release 3.2
(u'\ufa70', u'\ufad9'),  # CJK Compatibility Ideographs, release 4.1
(u'\u20000', u'\u2a6d6'),  # (UTF16) CJK Unified Ideographs Extension B, release 3.1
(u'\u2f800', u'\u2fa1d'),  # (UTF16) CJK Compatibility Supplement, release 3.1
(u'\uff00', u'\uffef'),  # Full width ASCII, full width of English punctuation, half width Katakana, half wide half width kana, Korean alphabet
(u'\u2e80', u'\u2eff'),  # CJK Radicals Supplement
(u'\u3000', u'\u303f'),  # CJK punctuation mark
(u'\u31c0', u'\u31ef'),  # CJK stroke  
(u'\u2f00', u'\u2fdf'),  # Kangxi Radicals
(u'\u2ff0', u'\u2fff'),  # Chinese character structure
(u'\u3100', u'\u312f'),  # Phonetic symbols
(u'\u31a0', u'\u31bf'),  # Phonetic symbols (Taiwanese and Hakka expansion)
(u'\ufe10', u'\ufe1f'),
(u'\ufe30', u'\ufe4f'),
(u'\u2600', u'\u26ff'),
(u'\u2700', u'\u27bf'),
(u'\u3200', u'\u32ff'),
(u'\u3300', u'\u33ff')
]

def isChineseChar(uchar):
    for start, end in chinese_charset:
        if start <= uchar <= end:
            return turn
    return false
```

### LINKED ISSUE #27: (minor) factor tokenizer code
Hi, there are three tokenizers, and at least two of them (`tokenize_13a()` and `tokenize_zh()`) contain the same processing steps for tokenizing Western-script characters. The code is not even identically written, although the regexes seem identical in the end. Code duplication is dangerous, in case future modifications are ever made to this code. I suggest to factor out this code.

### LINKED ISSUE #85: refactor tokenization classes
Tokenization is all hacked together with a dictionary of functions. This has caused problems with the Mecab tokenizer. It would be good to factor this out into a simple class hierarchy.

### PR DESCRIPTION
**Pull request by ozancaglayan**
- Added Multi30k multimodal MT test set metadata
  - Refactored all tokenizers into respective classes (fixes #85)
  - Refactored all metrics into respective classes
  - Moved utility functions into `utils.py`
  - Implemented signatures using `BLEUSignature` and `CHRFSignature` classes
  - Simplified checking of Chinese characters (fixes #5)
  - Unified common regexp tokenization codes for tokenizers (fixes #27)
  - Fixed --detail failing when no test sets are provided
  - Fixed multi-reference BLEU failing when tab-delimited reference stream is used
  - Removed lowercase option for ChrF which was not functional (#85)
  - Simplified ChrF and used the same I/O logic as BLEU to allow for future
    multi-reference reading
  - Added score regression tests for chrF using reference chrF++ implementation
  - Added multi-reference & tokenizer & signature tests

All new tests are added/verified on v1.4.10 and then tested again using this WIP branch to ensure that the refactorization did not break the tokenizers/metrics.

### CODE REVIEW

**Comment by martinpopel** [comment]
Hi @mjpost. I guess you are busy, but could you find time for reviewing this PR (and ideally #78 as well)?
I have further PR ideas and would like to know which branch I should work with.

**Comment by ozancaglayan** [comment]
#78 will be kind of stale if this branch gets merged. Then my initial plan
was to add an object oriented API for significance testing that can consume
arbitrary statistics both for bleu and chrf.

**Comment by mjpost** [comment]
Will do my best to review today! Thanks for your work here!

File: sacrebleu/dataset.py [hunk 1]
```diff
@@ -663,6 +663,31 @@
...
+        'en-fr': ['test_2018_flickr.en'],
+        'en-de': ['test_2018_flickr.en'],
+    },
 }
```
[the remark this answers is not in the document]
Comment (mjpost) [reply sacrebleu/dataset.py:691]: Also, doesn't have to be done here, but at some point it'd be nice to store the BibTeXs in plain-text files, and just load them...

**Comment by mjpost** [comment]
@martinpopel and @ozancaglayan, do you have any interest in being collaborators on this Github project? I could add you and then move to a model where any merges to `master` require someone to approve them.

File: sacrebleu/__init__.py [hunk 2]
```diff
@@ -17,10 +17,19 @@
 __version__ = '1.4.10'
 __description__ = 'Hassle-free computation of shareable, comparable, and reproducible BLEU scores'
```
[the remark this answers is not in the document]
Comment (ozancaglayan) [reply sacrebleu/__init__.py:19]: 1.4.11 right?

File: sacrebleu/compat.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2017--2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
```
[the remark this answers is not in the document]
Comment (ozancaglayan) [reply sacrebleu/compat.py:4]: thanks, I'll take a look at this.

**Comment by ozancaglayan** [comment]
I am interested esp. if I'll be able to close issues :p

**Comment by mjpost** [comment]
Before merging, can you also replace `sed -r` with `perl -pe` in `test.sh`? The `-r` flag is only available for GNU sed (and not on my Mac), but Perl should work everywhere.

File: sacrebleu/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjpost) [reply sacrebleu/__init__.py:19]: Yep. This is a big commit, but nothing really changes. We could do 1.5.0 to reflect it, but I think 1.4.11 is fine. I like understatement.

[the remark this answers is not in the document]
Comment (ozancaglayan) [reply sacrebleu/__init__.py:19]: I should have bumped this at the very first commit right for correctness? Do you care about this or should I just do it afterwards?

[the remark this answers is not in the document]
Comment (mjpost) [reply sacrebleu/__init__.py:19]: I don’t think the order of commits matters. When we merge we should do a squash merge anyway so it won’t matter.

[the remark this answers is not in the document]
Comment (ozancaglayan) [reply sacrebleu/__init__.py:19]: oh okay then!

File: sacrebleu/metrics/base.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        for name in sorted(self._sig.keys()):
+            value = self._sig[name]
+            if value is not None:
+                final_name = self._abbr[name] if self.short else name
+                pairs.append('{}.{}'.format(final_name, value))
+
+        return '+'.join(pairs)
+
+    def __repr__(self):
+        return self.__str__()
```
[the remark this answers is not in the document]
Comment (ozancaglayan) [reply sacrebleu/metrics/base.py:56]: I changed the `_sig` attribute to `info` so that the dictionary is public-facing.

**Comment by ozancaglayan** [comment]
I am done. Let's see whether pinning resolved mecab issue for Travis as well

File: sacrebleu/metrics/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjpost) [reply sacrebleu/metrics/base.py:56]: Great!

**Comment by mjpost** [comment]
Okay, almost ready to merge, then. Can you comment on the simplification for Chinese character checking (#5)? Does it compute the same result as before?

The best approach here would be to use the `regexp` library, which allows regex checking against Unicode character classes, e.g.,:

```python
def segment_char(line):
    # Chinese
    line = regex.sub(r'(\p{Han})', r' \1 ', line)
    # Korean
    line = regex.sub(r'(\p{Hangul})', r' \1 ', line)
    # Japanese
    line = regex.sub(r'(\p{Hiragana})', r' \1 ', line)
    line = regex.sub(r'(\p{Katakana})', r' \1 ', line)
    return line
```

But I don't think this would compute the same result as the existing code.

**Comment by ozancaglayan** [comment]
I just added a for loop instead of if-else's there and used a list of tuple
boundaries, so that should just be a refactoring without impact. I compared
test.sh outputs to 1.4.10 as well.

**Comment by mjpost** [comment]
@ozancaglayan, do you want to merge this, so that you're credited for it in the git commit? Please squash-merge and consolidate the commits into a summary (I suggest the following, based on your summary in the CHANGELOG):

```text
* Added Multi30k multimodal MT test set metadata
* Refactored all tokenizers into respective classes (fixes #85)
* Refactored all metrics into respective classes
* Moved utility functions into utils.py
* Implemented signatures using BLEUSignature and CHRFSignature classes, expose `Signature().info`
* Simplified checking of Chinese characters (fixes #5)
* Unified common regexp tokenization codes for tokenizers (fixes #27)
* Fixed --detail failing when no test sets are provided
* Fixed multi-reference BLEU failing when tab-delimited reference stream is used
* Removed lowercase option for ChrF which was not functional (#85)
* Simplified ChrF and used the same I/O logic as BLEU to allow for future
   multi-reference reading
* Added score regression tests for chrF using reference chrF++ implementation
* Added multi-reference & tokenizer & signature tests
* Pin mecab version to 0.996.5 as the newer ones are incompatible (fixes #94)
```

**Comment by martinpopel** [comment]
I would slightly prefer to not squash all the commits into one because they contain helpful logs and this PR changed a lot of code. If someone e.g. bisects a given issue it may be helpful to see the motivation for a given change. I am not sure whether the individual commits will be preserved somewhere if https://github.com/ozancaglayan/sacrebleu/tree/ozan-wip is squash-merged and deleted.
I see that some of the commits did not pass Travis, so that may be an obstacle for bisecting, but there are other commits in sacreBLEU's master which don't pass the Travis tests, so the obstacle is already there.

**Comment by mjpost** [comment]
Squash merging makes git histories much easier to read; each commit becomes one logical change.

Github preserves the history and all the intermediate commits even if the branch is deleted, so nothing is lost. Here's [an example from the Anthology](https://github.com/acl-org/acl-anthology/pull/831).

**Comment by martinpopel** [comment]
I know squashing and its benefits (I use it sometimes locally and in that case the intermediate commits are completely lost). 

> Github preserves the history and all the intermediate commits even if the branch is deleted, so nothing is lost. 

In that case it is OK to squash merge.

**Comment by mjpost** [comment]
Thanks, all! I tagged and released version 1.4.11.

**Comment by fhieber** [comment]
```bash
> pip install --upgrade sacrebleu
[...]
Successfully installed sacrebleu-1.4.11
> sacrebleu
Traceback (most recent call last):
  File "/Users/fhieber/miniconda3/bin/sacrebleu", line 5, in <module>
    from sacrebleu.sacrebleu import main
  File "/Users/fhieber/miniconda3/lib/python3.6/site-packages/sacrebleu/__init__.py", line 25, in <module>
    from .tokenizers import TOKENIZERS, DEFAULT_TOKENIZER
ModuleNotFoundError: No module named 'sacrebleu.tokenizers'
```

**Comment by martinpopel** [comment]
@fhieber Thanks for reporting the bug. See #96 and #97. @mjpost should upload a new version to PyPI.

**Comment by mjpost** [comment]
@fhieber finding my bugs like old times

**Pull request merged**